projects/gloo/pkg/utils/names.go

```
@@ -0,0 +1,26 @@
+package utils
+
+import (
+	"fmt"
+
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+)
+
+func RouteConfigName(listener *v1.Listener) string {
+	return listener.GetName() + "-routes"
+}
+
+func MatchedRouteConfigName(listener *v1.Listener, matcher *v1.Matcher) string {
+	hybridListener := listener.GetHybridListener()
+	if hybridListener == nil {
+		return RouteConfigName(listener)
+	}
+
+	for i, mg := range hybridListener.GetMatchedListeners() {
+		if mg.GetMatcher().Equal(matcher) {
+			return fmt.Sprintf("%s-%d", RouteConfigName(listener), i)
+		}
+	}
+
+	return fmt.Sprintf("%s-%s", RouteConfigName(listener), matcher.String())
```

I don't quite follow why we iterate the matchers of the gateways and then if we don't find one return the stringified matcher. Can't we just return the routeConfigName + stringified matcher without having to iterate the list?

---

The goal here is to have a deterministic, unique, and ideally concise route config name
`matcher.String()` does not satisfy this since it does not account for slice/map ordering and is extremely verbose
Using the index of the matcher's MatchedListener within the HybridListener satisfies all three

We really should never exit the `for` loop. The catchall case is mainly there as an "emergency" measure, to output something unique so we can be sure that nothing is overwritten.

---

That makes sense. Since matcher.String does not satisfy the requirements however, should we instead return an error? I'm concerned that we'll lose the context on why we have this return statement.

Also, how does the slice ordering affect matcher.String? I'm confused how adding the index differs from the string value. The index will be unique but will it be deterministic?

---

The `.String()` function returns a full representation of the struct, which has no bound on length...given that the name is propagated to Envoy, used in testing, and potentially visible elsewhere (I haven't followed that thread all the way to the end) it's definitely desirable, if not strictly necessary from a technical standpoint, to have a concise, ideally constant sized, name

I'm not sure if the slice ordering piece would affect determinism in practice, since references should always stem from the same source of truth (ie the Gateway CR), but it is the case that structs that has the same values in a slice field but in a different order will have different `.String()` outputs.

Likewise, the index should be deterministic since all references stem from the Gateway CR

(If you've got more questions about how/why to use index here perhaps reach out to the [developer who suggested it](https://github.com/solo-io/gloo/pull/5602#pullrequestreview-813061899) 😉)

---

As far as returning an error, it's mostly just clunky, as we use the result directly to index into maps within reports and set fields directly in struct construction, so if we can prove/be super highly confident that it will never happen in practice I'd prefer not to
That said, no technical reason it can't be done